api/apiclient.go

```diff
@@ -367,8 +368,12 @@ func (st *state) connectStream(path string, attrs url.Values, extraHeaders http.
 		TLSClientConfig: st.tlsConfig,
 	}
 	var requestHeader http.Header
-	if st.tag != "" {
-		requestHeader = jujuhttp.BasicAuthHeader(st.tag, st.password)
+	token, err := st.LoginToken()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if st.tag != "" || token != "" {
```

Comment:
I'm not a big fan of this logic but I believe it's needed because of the macaroon logic further below. The userpass login provider also handles macaroon login. During macaroon auth, the tag and password are both empty and I'm unsure what the server's behaviour is if the basic auth header is set and macaroons are included, there doesn't seem to be tests for the stream connector when using macaroons.

Reply:
Will this work? Now that the underlying provider can return ErrorTokenNotAvailable, doesn't the error check need to account for that? And we'd avoid the `token != ""` check since we don't want to special case ""
But, ErrorTokenNotAvailable seems to imply a transient error whereas the logic here seems to want to make a decision based on a static decision - there is a token or there is not. We've introduced the error in this PR and nowhere is using it. So what's the plan for it?

Reply:
I'll make this change because I think it's the right thing regarding the contract that `LoginToken` is trying to form.